File: tests/python/core/schema/test_joint_se3_pose.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Unit tests for JointSe3PoseOutput in isaacteleop.schema.

JointSe3PoseOutput is a FlatBuffers table for a sparse set of tracked joint poses:
- joints: JointSe3Pose entries keyed by JointName, sorted ascending on the wire
- device_id: stable identity of the producing device

A joint absent from joints is not tracked; there is no per-joint validity flag.
Timestamps are carried by JointSe3PoseOutputRecord, not JointSe3PoseOutput.

Note: Python code should only READ this data (created by C++ trackers), not modify it.
"""

import pytest

from isaacteleop.schema import (
    DeviceDataTimestamp,
    JointName,
    JointType,
    JointSe3Pose,
    JointSe3PoseOutput,
    JointSe3PoseOutputRecord,
    Point,
    Pose,
    Quaternion,
)

# Every tip Manus produces, thumb→pinky.
TIPS = [
    JointName.HAND_RAW_THUMB_TIP,
    JointName.HAND_RAW_INDEX_TIP,
    JointName.HAND_RAW_MIDDLE_TIP,
    JointName.HAND_RAW_RING_TIP,
    JointName.HAND_RAW_LITTLE_TIP,
]


def tip(joint: JointName, x: float) -> JointSe3Pose:
    """One tip at (x, 0, 0), so a position identifies which entry it came from."""
    return JointSe3Pose(joint, Pose(Point(x, 0.0, 0.0), Quaternion(0.0, 0.0, 0.0, 1.0)))


class TestJointSe3Pose:
    """Tests for the JointSe3Pose struct."""

    def test_default_construction(self):
        """Default construction yields the UNKNOWN key and a zero pose."""
        joint = JointSe3Pose()

        assert joint.joint == JointName.UNKNOWN
        assert joint.pose.position.x == pytest.approx(0.0)

    def test_parameterized_construction(self):
        """Construction carries the key and the pose through unchanged."""
        joint = JointSe3Pose(
            JointName.HAND_RAW_THUMB_TIP,
            Pose(Point(1.0, 2.0, 3.0), Quaternion(0.0, 0.0, 0.0, 1.0)),
        )

        assert joint.joint == JointName.HAND_RAW_THUMB_TIP
        assert joint.pose.position.x == pytest.approx(1.0)
        assert joint.pose.position.y == pytest.approx(2.0)
        assert joint.pose.position.z == pytest.approx(3.0)
        assert joint.pose.orientation.w == pytest.approx(1.0)

    def test_repr(self):
        """__repr__ names the joint rather than printing its ordinal."""
        joint = tip(JointName.HAND_RAW_THUMB_TIP, 1.0)

        assert "HAND_RAW_THUMB_TIP" in repr(joint)

    def test_joint_names_are_in_their_allocated_block(self):
        """Every tip sits in the JointType.HAND_RAW block, which is what lets a writer
        stamp type and a reader check it against the keys."""
        assert int(JointName.UNKNOWN) == 0
        assert int(JointType.UNKNOWN) == 0
        for joint in TIPS:
            assert int(JointType.HAND_RAW) <= int(joint) < 1200
            assert int(joint) > int(JointType.HAND_OPENXR)


class TestJointSe3PoseOutput:
    """Tests for the per-frame output table."""

    def test_joints_are_sorted_on_the_wire(self):
        """Joints given out of order come back sorted ascending.

        LookupByKey binary-searches, so an unsorted vector would return the wrong joint
        rather than fail. The binding sorts on the way in to make that unreachable.
        """
        out = JointSe3PoseOutput(
            type=JointType.HAND_RAW,
            joints=[
                tip(JointName.HAND_RAW_LITTLE_TIP, 4.0),
                tip(JointName.HAND_RAW_THUMB_TIP, 0.0),
                tip(JointName.HAND_RAW_MIDDLE_TIP, 2.0),
            ],
            device_id="manus_sensors_left",
        )

        keys = [int(j.joint) for j in out.joints]
        assert keys == sorted(keys)

    def test_lookup_returns_the_matching_joint(self):
        """Each written joint is found by name, carrying its own pose."""
        out = JointSe3PoseOutput(
            type=JointType.HAND_RAW,
            joints=[tip(joint, float(i)) for i, joint in enumerate(TIPS)],
            device_id="manus_sensors_left",
        )

        assert out.type == JointType.HAND_RAW
        assert out.device_id == "manus_sensors_left"
        assert len(out.joints) == len(TIPS)
        for i, joint in enumerate(TIPS):
            found = out.lookup(joint)
            assert found is not None
            assert found.joint == joint
            # Checking the position, not just non-None, is what proves the lookup
            # landed on the right entry rather than a neighbour.
            assert found.pose.position.x == pytest.approx(float(i))

    def test_lookup_of_an_untracked_joint_returns_none(self):
        """A joint absent from the frame misses instead of returning a neighbour."""
        out = JointSe3PoseOutput(
            type=JointType.HAND_RAW,
            joints=[
                tip(JointName.HAND_RAW_MIDDLE_TIP, 2.0),
                tip(JointName.HAND_RAW_LITTLE_TIP, 4.0),
            ],
            device_id="manus_sensors_left",
        )

        # RING sits between the two written keys, so a lookup that fell through to the
        # nearest entry would answer with one of them.
        assert out.lookup(JointName.HAND_RAW_RING_TIP) is None
        assert out.lookup(JointName.UNKNOWN) is None

    def test_empty_output(self):
        """A device reporting nothing yields no joints and no lookups."""
        out = JointSe3PoseOutput()

        assert out.type == JointType.UNKNOWN
        assert len(out.joints) == 0
        assert out.lookup(JointName.HAND_RAW_THUMB_TIP) is None

    def test_untracked_frame_still_names_its_family(self):
        """An all-untracked frame keeps its type.

        This is why type is a field and not derived from the keys: with no keys there is
        nothing to derive it from.
        """
        out = JointSe3PoseOutput(
            type=JointType.HAND_RAW, device_id="manus_sensors_left"
        )

        assert out.type == JointType.HAND_RAW
        assert len(out.joints) == 0

    def test_duplicate_joints_are_rejected(self):
        """Two entries under one key make lookup ambiguous, so construction fails.

        Sorting cannot fix this: LookupByKey binary-searches and cannot say which of two
        equal keys it landed on.
        """
        with pytest.raises(ValueError, match="HAND_RAW_INDEX_TIP"):
            JointSe3PoseOutput(
                type=JointType.HAND_RAW,
                joints=[
                    tip(JointName.HAND_RAW_INDEX_TIP, 1.0),
                    tip(JointName.HAND_RAW_INDEX_TIP, 2.0),
                ],
                device_id="manus_sensors_left",
            )

    def test_joints_outside_the_declared_type_are_rejected(self):
        """A key from another family contradicts type, which one frame cannot do."""
        with pytest.raises(ValueError, match="HAND_RAW_THUMB_TIP"):
            JointSe3PoseOutput(
                type=JointType.HAND_OPENXR,
                joints=[tip(JointName.HAND_RAW_THUMB_TIP, 0.0)],
                device_id="manus_sensors_left",
            )

    def test_untyped_frame_with_joints_is_rejected(self):
        """UNKNOWN names no family, so keys with nothing to name them is a writer bug.

        The empty case stays legal: that is the one frame UNKNOWN fits.
        """
        with pytest.raises(ValueError, match="UNKNOWN"):
            JointSe3PoseOutput(
                joints=[tip(JointName.HAND_RAW_THUMB_TIP, 0.0)],
                device_id="manus_sensors_left",
            )

    def test_repr(self):
        """__repr__ includes the device id and the joint count."""
        out = JointSe3PoseOutput(
            type=JointType.HAND_RAW,
            joints=[tip(JointName.HAND_RAW_THUMB_TIP, 0.0)],
            device_id="manus_sensors_right",
        )

        repr_str = repr(out)
        assert "manus_sensors_right" in repr_str
        assert "joints=1" in repr_str
        assert "HAND_RAW" in repr_str


class TestJointSe3PoseOutputRecord:
    """Tests for the MCAP recording wrapper."""

    def test_construction_with_timestamp(self):
        """The record carries DeviceDataTimestamp alongside the payload."""
        out = JointSe3PoseOutput(
            type=JointType.HAND_RAW,
            joints=[tip(JointName.HAND_RAW_THUMB_TIP, 1.0)],
            device_id="manus_sensors_left",
        )
        record = JointSe3PoseOutputRecord(
            out, DeviceDataTimestamp(1000000000, 2000000000, 3000000000)
        )

        assert record.timestamp.available_time_local_common_clock == 1000000000
        assert record.timestamp.sample_time_local_common_clock == 2000000000
        assert record.timestamp.sample_time_raw_device_clock == 3000000000
        assert record.data.device_id == "manus_sensors_left"

    def test_payload_less_record(self):
        """A record may carry a timestamp and no payload: MCAP's frame sentinel."""
        record = JointSe3PoseOutputRecord(None, DeviceDataTimestamp(1, 2, 3))

        assert record.data is None
        assert record.timestamp.available_time_local_common_clock == 1

File: tests/cpp/core/schema/test_joint_se3_pose.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

// Unit tests for the generated JointSe3PoseOutput FlatBuffer message.

#include <catch2/catch_test_macros.hpp>
#include <flatbuffers/flatbuffers.h>

// Include generated FlatBuffer headers.
#include <schema/joint_se3_pose_family.hpp>
#include <schema/joint_se3_pose_generated.h>
#include <schema/timestamp_generated.h>

#include <string>
#include <type_traits>
#include <vector>

// =============================================================================
// Compile-time verification of FlatBuffer field IDs.
// These ensure schema field IDs remain stable across changes.
// VT values are computed as: (field_id + 2) * 2.
// =============================================================================
#define VT(field) (field + 2) * 2
static_assert(core::JointSe3PoseOutput::VT_TYPE == VT(0));
static_assert(core::JointSe3PoseOutput::VT_JOINTS == VT(1));
static_assert(core::JointSe3PoseOutput::VT_DEVICE_ID == VT(2));

static_assert(core::JointSe3PoseOutputRecord::VT_DATA == VT(0));
static_assert(core::JointSe3PoseOutputRecord::VT_TIMESTAMP == VT(1));

// =============================================================================
// Compile-time verification of FlatBuffer field types.
// =============================================================================
static_assert(std::is_same_v<decltype(std::declval<core::JointSe3Pose>().joint()), core::JointName>);
static_assert(std::is_same_v<decltype(std::declval<core::JointSe3Pose>().pose()), const core::Pose&>);

// A struct is stored inline with no vtable, so its size is baked into every layout enclosing
// it. Resizing JointSe3Pose silently misreads every recorded vector of them.
static_assert(sizeof(core::JointSe3Pose) == 32);

namespace
{

// One tip at (index, 0, 0), so a joint's position identifies which entry it came from.
core::JointSe3Pose tip(core::JointName joint, float x)
{
    return core::JointSe3Pose(joint, core::Pose(core::Point(x, 0.0F, 0.0F), core::Quaternion(0.0F, 0.0F, 0.0F, 1.0F)));
}

} // namespace

TEST_CASE("JointSe3PoseOutput looks joints up by name", "[joint_se3_pose]")
{
    // Deliberately unsorted, and missing RING: the vector is sparse, and a producer must not
    // have to hand the builder its joints in order.
    std::vector<core::JointSe3Pose> joints{
        tip(core::JointName_HAND_RAW_LITTLE_TIP, 4.0F),
        tip(core::JointName_HAND_RAW_THUMB_TIP, 0.0F),
        tip(core::JointName_HAND_RAW_MIDDLE_TIP, 2.0F),
        tip(core::JointName_HAND_RAW_INDEX_TIP, 1.0F),
    };

    flatbuffers::FlatBufferBuilder builder(1024);
    const auto device_id = builder.CreateString("manus_sensors_left");
    const auto joints_offset = builder.CreateVectorOfSortedStructs(&joints);
    builder.Finish(core::CreateJointSe3PoseOutput(builder, core::JointType_HAND_RAW, joints_offset, device_id));

    const auto* out = flatbuffers::GetRoot<core::JointSe3PoseOutput>(builder.GetBufferPointer());
    REQUIRE(out->joints() != nullptr);
    REQUIRE(out->joints()->size() == 4);
    CHECK(out->device_id()->str() == "manus_sensors_left");
    CHECK(out->type() == core::JointType_HAND_RAW);

    SECTION("every written joint comes back, carrying its own pose")
    {
        // LookupByKey on an unsorted vector returns a neighbour rather than failing, so
        // checking the position -- not just non-null -- is what proves the sort happened.
        const std::vector<std::pair<core::JointName, float>> expected{
            { core::JointName_HAND_RAW_THUMB_TIP, 0.0F },
            { core::JointName_HAND_RAW_INDEX_TIP, 1.0F },
            { core::JointName_HAND_RAW_MIDDLE_TIP, 2.0F },
            { core::JointName_HAND_RAW_LITTLE_TIP, 4.0F },
        };
        for (const auto& [joint, x] : expected)
        {
            const auto* found = out->joints()->LookupByKey(joint);
            REQUIRE(found != nullptr);
            CHECK(found->joint() == joint);
            CHECK(found->pose().position().x() == x);
        }
    }

    SECTION("the vector is sorted ascending on the wire")
    {
        core::JointName previous = core::JointName_UNKNOWN;
        for (const auto* joint : *out->joints())
        {
            CHECK(joint->joint() > previous);
            previous = joint->joint();
        }
    }

    SECTION("an untracked joint misses rather than returning a neighbour")
    {
        // RING sits between MIDDLE and LITTLE, so a lookup that fell through to the nearest
        // entry would answer with one of those.
        CHECK(out->joints()->LookupByKey(core::JointName_HAND_RAW_RING_TIP) == nullptr);
        CHECK(out->joints()->LookupByKey(core::JointName_UNKNOWN) == nullptr);
    }
}

TEST_CASE("JointSe3PoseOutput carries no joints when the device reports none", "[joint_se3_pose]")
{
    flatbuffers::FlatBufferBuilder builder(1024);
    std::vector<core::JointSe3Pose> joints;
    const auto joints_offset = builder.CreateVectorOfSortedStructs(&joints);
    builder.Finish(
        core::CreateJointSe3PoseOutput(builder, core::JointType_HAND_RAW, joints_offset, builder.CreateString("")));

    const auto* out = flatbuffers::GetRoot<core::JointSe3PoseOutput>(builder.GetBufferPointer());
    REQUIRE(out->joints() != nullptr);
    CHECK(out->joints()->size() == 0);
    // The point of carrying type as a field: an all-untracked frame still names its family,
    // which no amount of inspecting the (empty) key vector could tell you.
    CHECK(out->type() == core::JointType_HAND_RAW);
    CHECK(out->joints()->LookupByKey(core::JointName_HAND_RAW_THUMB_TIP) == nullptr);
}

TEST_CASE("JointSe3PoseOutputT round-trips through the object API", "[joint_se3_pose][native]")
{
    core::JointSe3PoseOutputT native;
    native.type = core::JointType_HAND_RAW;
    native.device_id = "manus_sensors_right";
    native.joints = {
        tip(core::JointName_HAND_RAW_THUMB_TIP, 0.0F),
        tip(core::JointName_HAND_RAW_LITTLE_TIP, 4.0F),
    };

    flatbuffers::FlatBufferBuilder builder(1024);
    builder.Finish(core::JointSe3PoseOutput::Pack(builder, &native));

    const auto* out = flatbuffers::GetRoot<core::JointSe3PoseOutput>(builder.GetBufferPointer());
    CHECK(out->device_id()->str() == "manus_sensors_right");
    CHECK(out->type() == core::JointType_HAND_RAW);
    REQUIRE(out->joints()->size() == 2);
    // Pack() writes the vector in the order given (CreateVectorOfStructs, not the sorted
    // variant), so callers that build through it must sort first. This pins that behaviour:
    // if flatc ever starts sorting here, the writers that sort by hand need revisiting.
    CHECK((*out->joints())[0]->joint() == core::JointName_HAND_RAW_THUMB_TIP);
    CHECK((*out->joints())[1]->joint() == core::JointName_HAND_RAW_LITTLE_TIP);
}

TEST_CASE("JointName values fall in their JointType block", "[joint_se3_pose]")
{
    // The family of a joint is the greatest JointType not exceeding it. Block widths differ
    // (100 for hands, 1000 for bodies), so rounding to a fixed width is not the rule; this
    // pins the allocation the writers rely on when they stamp `type`.
    for (const auto joint :
         { core::JointName_HAND_RAW_THUMB_TIP, core::JointName_HAND_RAW_INDEX_TIP, core::JointName_HAND_RAW_MIDDLE_TIP,
           core::JointName_HAND_RAW_RING_TIP, core::JointName_HAND_RAW_LITTLE_TIP })
    {
        CHECK(static_cast<int>(joint) >= static_cast<int>(core::JointType_HAND_RAW));
        // Below the next allocated base, so these never read as HAND_OPENXR or a body family.
        CHECK(static_cast<int>(joint) < 1200);
        CHECK(static_cast<int>(joint) > static_cast<int>(core::JointType_HAND_OPENXR));
    }
    CHECK(static_cast<int>(core::JointType_UNKNOWN) == 0);
}

TEST_CASE("joint_family maps a joint onto its block base", "[joint_se3_pose]")
{
    CHECK(core::joint_family(core::JointName_HAND_RAW_THUMB_TIP) == core::JointType_HAND_RAW);
    CHECK(core::joint_family(core::JointName_HAND_RAW_LITTLE_TIP) == core::JointType_HAND_RAW);
    // Below every allocated base.
    CHECK(core::joint_family(core::JointName_UNKNOWN) == core::JointType_UNKNOWN);
    // A value inside the HAND_OPENXR block that has no enumerator yet still reads as that family:
    // the rule is the block it falls in, not the names allocated so far.
    CHECK(core::joint_family(static_cast<core::JointName>(1042)) == core::JointType_HAND_OPENXR);
}
